dali/imgcodec/parsers/jpeg.cc

```diff
@@ -13,18 +13,70 @@
 // limitations under the License.
 
 #include "dali/imgcodec/parsers/jpeg.h"
+#include "dali/core/byte_io.h"
 
 namespace dali {
 namespace imgcodec {
 
+using jpeg_marker_t = std::array<uint8_t, 2>;
+
+constexpr jpeg_marker_t sos_marker = {0xff, 0xda};
+constexpr jpeg_marker_t soi_marker = {0xff, 0xd8};
+constexpr jpeg_marker_t eoi_marker = {0xff, 0xd9};
+
+bool IsValidMarker(const jpeg_marker_t &marker) {
+  return marker[0] == 0xff && marker[1] != 0x00;
+}
+
+bool IsSofMarker(const jpeg_marker_t &marker) {
+  // According to https://www.w3.org/Graphics/JPEG/itu-t81.pdf table B.1 Marker code assignments
+  // SOF markers are from range 0xFFC0-0xFFCF, excluding 0xFFC4, 0xFFC8 and 0xFFCC.
+  if (!IsValidMarker(marker) || marker[1] < 0xc0 || marker[1] > 0xcf) return false;
+  return marker[1] != 0xc4 && marker[1] != 0xc8 && marker[1] != 0xcc;
+}
+
 ImageInfo JpegParser::GetInfo(ImageSource *encoded) const {
-  ImageInfo info;
-  // info.shape = {h, w, c};   // TODO(janton)
+  ImageInfo info{};
+  auto stream = encoded->Open();
+
+  jpeg_marker_t first_marker = stream->ReadOne<jpeg_marker_t>();
+  DALI_ENFORCE(first_marker == soi_marker);
+
+  bool read_shape = false;
+  while (!read_shape) {
+    jpeg_marker_t marker;
+    marker[0] = stream->ReadOne<uint8_t>();
+    DALI_ENFORCE(marker[0] == 0xff, make_string("A valid marker must start with 0xFF"));
+    // https://www.w3.org/Graphics/JPEG/itu-t81.pdf section B.1.1.2 Markers
+    // Any marker may optionally be preceded by any number of fill bytes, 
+    // which are bytes assigned code X’FF’
+    do {
+      marker[1] = stream->ReadOne<uint8_t>();
+    } while (marker[1] == 0xff);
+    DALI_ENFORCE(IsValidMarker(marker), make_string("Invalid marker"));
+    if (marker == sos_marker) break;
```

Comment:
Nitpick:
```suggestion
    if (marker == sos_marker)
      break;
```
I'm not a fan of braces around single statements, but putting the `if` contents in a new line makes this impossible:
```C++
    if (marker == sos_marker) break;
    {
       // looks like `if` body, but it's just a scope
    }
```

Reply:
Fixed